syntax = "proto3";

package yandex.cloud.audit.lockbox;

import "google/protobuf/struct.proto";
import "google/protobuf/timestamp.proto";
import "google/rpc/status.proto";
import "yandex/cloud/audit/common.proto";
import "yandex/cloud/validation.proto";

option go_package = "a.yandex-team.ru/cloud/bitbucket/public-api/yandex/cloud/audit/lockbox;lockbox";
option java_package = "yandex.cloud.api.audit.lockbox";
option java_outer_classname = "SecretEvents";

message CreateSecret {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    string secret_version_id = 3;
    string kms_key_id = 4;
    bool deletion_protection = 5;
    string status = 6;
    string secret_version_status = 7;
    repeated string secret_version_payload_entry_keys = 8;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message UpdateSecret {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    string secret_version_id = 3;
    string kms_key_id = 4;
    bool deletion_protection = 5;
    string status = 6;
    string secret_version_status = 7;
    repeated string secret_version_payload_entry_keys = 8;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message DeleteSecret {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    string secret_version_id = 3;
    string kms_key_id = 4;
    bool deletion_protection = 5;
    string status = 6;
    string secret_version_status = 7;
    repeated string secret_version_payload_entry_keys = 8;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message ActivateSecret {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    string secret_version_id = 3;
    string kms_key_id = 4;
    bool deletion_protection = 5;
    string status = 6;
    string secret_version_status = 7;
    repeated string secret_version_payload_entry_keys = 8;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message DeactivateSecret {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    string secret_version_id = 3;
    string kms_key_id = 4;
    bool deletion_protection = 5;
    string status = 6;
    string secret_version_status = 7;
    repeated string secret_version_payload_entry_keys = 8;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message AddVersion {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    repeated string payload_entry_keys = 3;
    string new_secret_version_id = 4;
    string old_secret_version_id = 5;
    repeated PayloadEntryDelta payload_entry_deltas = 6;
    string status = 7;
    string secret_version_status = 8;


    message PayloadEntryDelta {
      string key = 1;
      DeltaType delta_type = 2;

      enum DeltaType {
        DELTA_TYPE_UNSPECIFIED = 0;
        ADDED = 1;
        UPDATED = 2;
        REMOVED = 3;
      }
    }
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message ScheduleVersionDestruction {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    string secret_version_id = 3;
    google.protobuf.Timestamp destroy_at = 4;
    string status = 5;
    string secret_version_status = 6;
    repeated string secret_version_payload_entry_keys = 7;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message CancelVersionDestruction {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    string secret_version_id = 3;
    string secret_version_status = 4;
    repeated string secret_version_payload_entry_keys = 5;
    string status = 6;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message SetSecretAccessBindings {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    repeated audit.AccessBinding access_bindings = 3;
    repeated audit.AccessBindingDelta access_binding_deltas = 4;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}

message UpdateSecretAccessBindings {
  string event_id = 1 [(required) = true];
  string event_source = 2 [(required) = true];
  string event_type = 3 [(required) = true];
  google.protobuf.Timestamp event_time = 4 [(required) = true];
  Authentication authentication = 5 [(required) = true];
  Authorization authorization = 6 [(required) = true];
  ResourceMetadata resource_metadata = 7 [(required) = true];
  RequestMetadata request_metadata = 8 [(required) = true];
  EventStatus event_status = 9 [(required) = true];
  google.rpc.Status error = 10;
  EventDetails details = 11 [(required) = true];
  google.protobuf.Struct request_parameters = 12;
  google.protobuf.Struct response = 13;

  message EventDetails {
    string secret_id = 1;
    string secret_name = 2;
    repeated audit.AccessBindingDelta access_binding_deltas = 3;
    repeated audit.AccessBinding access_bindings = 4;
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}
